tests/test_utils_dict_support.py

```diff
@@ -89,6 +89,64 @@ def test_merging_dictionaries_with_lists():
     assert_semantically_equal(return_d, {'a': 1, 'b': [3, 6, 5], 'c': [1, 8, 90]})
 
 
+def test_preserves_duplicates_in_input_list_when_merging_dictionaries():
+    return_d = dict_support.merge_complex_dictionaries(
+        {'b': [6, 6]},
+        {'b': [5]}
+    )
+    assert_semantically_equal(return_d, {'b': [6, 6, 5]})
+
+
+def test_preserves_resulting_duplicates_in_lists_when_merging_dictionaries():
+    return_d = dict_support.merge_complex_dictionaries(
+        {'b': [6]},
+        {'b': [5, 6]}
+    )
+    assert_semantically_equal(return_d, {'b': [6, 5, 6]})
+
+
+def test_preserves_lists_order_when_merging_dictionaries():
+    return_d = dict_support.merge_complex_dictionaries(
+        {'b': [1, 2]},
+        {'b': [3, 4]}
+    )
+    assert_semantically_equal(return_d, {'b': [1, 2, 3, 4]})
+
+
+def test_combines_sets_with_or_operator_when_merging_dictionaries():
+    return_d = dict_support.merge_complex_dictionaries(
+        {'b': {1, 2}},
+        {'b': {2, 3}}
+    )
+    assert_semantically_equal(return_d, {'b': {1, 2, 3}})
+
+
+def test_does_not_mutate_parameters_when_merging_complex_dictionaries():
+    A = {
+        'a': {1, 2},
+        'b': [6, 6],
+        'c': {
+            'a': {1, 2},
+            'b': [6, 6]
+        },
+        'd': 'hey'
+    }
+    B = {
+        'a': {2, 3},
+        'b': [5],
+        'c': {
+            'a': {2, 3},
+            'b': [5]
+        },
+        'd': 'yo'
+    }
+
+    return_d = dict_support.merge_complex_dictionaries(A, B)
+
+    assert_semantically_equal(A, {'a': {1, 2}, 'b': [6, 6], 'c': {'a': {1, 2}, 'b': [6, 6]}, 'd': 'hey'})
```

Comment:
Could you use deep copies of the two dicts taken before the call to merge as your "befores" and then compare `A` and `B` to their corresponding "before" states?